### ISSUE TITLE
Which objective functions does TensorFlow Probability provide?

### ISSUE THREAD

**Issue by nbro** [tensorflow/probability#740]
To train a probabilistic neural network (PNN), the ELBO loss is usually used. The ELBO loss is composed of the KL loss between the prior and the variational posterior distributions, for each layer, and the likelihood loss, which needs to be implemented by the programmer. However, it may be handy to have some default implementations of this likelihood loss (similarly to the TF's or Keras' implementations of the cross-entropy, MSE, etc., losses). Am I missing something? Why didn't you provide the default implementations of these losses (at least, the most commonly used ones)?

**Comment by brianwa84** [comment]
Typically the likelihood portion will be computed from the output of the
network against the labels, ie.

m = tf.keras.Sequential([
  ...,
  tfp.layers.IndependentBernoulli(..)
])
d = m(features)
nll = -d.log_prob(labels)

Is that what you mean?

**Comment by nbro** [comment]
@brianwa84 Yes, exactly, that's what I mean. Your `nll` could have a (simple) default implementation, so that it works with different distributions, etc.

**Comment by brianwa84** [comment]
I guess I'd be concerned that doing so would mask what's already a quite
simple one-liner behind a facade, making it more imposing to a reader. What
do you think the user code might look like in an ideal world?

Brian Patton | Software Engineer | bjp@google.com

**Comment by nbro** [comment]
@brianwa84 The Keras' or even the TF implementations of certain loss functions may also be simple (honestly, I haven't looked at their implementation yet).  For example, people could forget the minus in front of it. By providing a default implementation, these mistakes would not occur. Of course, Bayesian deep learning is still a relatively new field, so things may change meanwhile. 

In an ideal world, I would only need to specify the loss function that will be used to train the Bayesian NN, without having to implement it. I should only need to implement it in the case it is a custom loss.
